File: Day103_2025/Q103_Leetcode.py
class Solution:
    MOD = 10**9 + 7

    def countGoodNumbers(self, n: int) -> int:
        return (self.mod_pow(5, (n + 1) // 2) * self.mod_pow(4, n // 2)) % self.MOD

    def mod_pow(self, base: int, exp: int) -> int:
        result = 1
        base %= self.MOD
        while exp > 0:
            if exp % 2 == 1:
                result = (result * base) % self.MOD
            base = (base * base) % self.MOD
            exp //= 2
        return result

File: Day103_2025/Q103_Leetcode.java
class Q103_Leetcode {
    static final int MOD = 1000000007;

    public int countGoodNumbers(long n) {
        return (int) (modPow(5, (n + 1) / 2) * modPow(4, n / 2) % MOD);
    }

    private long modPow(long base, long exp) {
        long result = 1;
        base %= MOD;
        while (exp > 0) {
            if ((exp & 1) == 1)
                result = result * base % MOD;
            base = base * base % MOD;
            exp >>= 1;
        }
        return result;
    }
}
